lib/sqlitex/query.ex

```diff
@@ -0,0 +1,61 @@
+defmodule Sqlitex.Query do
+  use Pipe
+
+  def query(db, sql, opts \\ []) do
+    pipe_matching {:ok, _},
+        prepare(sql, db)
+        |> bind_values(opts)
+        |> execute(opts)
+  end
+
+  defp bind_values({:ok, statement}, opts) do
+    params = params_from_opts(opts)
+    case :esqlite3.bind(statement, params) do
+      {:error,_}=error -> error
+      :ok -> {:ok, statement}
+    end
+  end
+
+  defp datetime_to_string({{yr, mo, da}, {hr, mi, se, usecs}}) do
+    [zero_pad(yr, 4), "-", zero_pad(mo, 2), "-", zero_pad(da, 2), " ", zero_pad(hr, 2), ":", zero_pad(mi, 2), ":", zero_pad(se, 2), ".", zero_pad(usecs, 6)]
+    |> Enum.join
+  end
+
+  defp execute({:ok, statement}, opts) do
+    into = into_from_opts(opts)
+    types = :esqlite3.column_types(statement)
+    columns = :esqlite3.column_names(statement)
+    data = :esqlite3.fetchall(statement)
+    to_rows(types, columns, data, into)
```

Comment:
I got around the issue of having a different interface between `execute` and `to_rows` by making `execute` call directly into `to_rows`. That way the whole function either returns `{:ok, results}` or `{:error, some_error}`. Not a brilliant solution, but it works well enough.